projects/interfaces/main.go

```diff
@@ -0,0 +1,128 @@
+)
+
+// ---------- Implementing our own bytes.Buffer
+
+// the original Buffer type definition:
+//
+//	type Buffer struct {
+//		buf      []byte // contents are the bytes buf[off : len(buf)]
+//		off      int    // read at &buf[off], write at &buf[len(buf)]
+//		lastRead readOp // last read operation, so that Unread* can work correctly.
+//	}
+type OurByteBuffer struct {
+	buf []byte
+}
+
+// the original Buffer Bytes() method signature:
+// func (b *Buffer) Bytes() []byte
+func (b *OurByteBuffer) Bytes() []byte {
+	// simply return the buffer
+	return b.buf
+}
+
+// the original Buffer Write() method signature:
+// func (b *Buffer) Write(p []byte) (n int, err error)
+func (b *OurByteBuffer) Write(p []byte) (int, error) {
+	// Q: Why is no error returned from append(?)
+	// A: The append function in Go doesn't generally return an error because it performs a safe operation that cannot fail in most cases.
+	// The reason for this is related to how Go handles memory allocation and slice operations.
+
+	// append the byte slice values (passed to the function) to the buffer
+	b.buf = append(b.buf, p...)
+
+	// return the number of bytes that were appended, and an error... although it can never be not nil(??)
+	return len(p), nil
+}
+
+// the original Buffer Read() method signature:
+// func (b *Buffer) Read(p []byte) (n int, err error)
+func (b *OurByteBuffer) Read(p []byte) (int, error) {
+	// if the buffer is empty then return 0 and io.EOF
+	if len(b.buf) == 0 {
+		// Q: What is is.EOF ?
+		// A: EOF stands for End of File
+		// When a function returns io.EOF, it means that there's no more data to be read from the input source.
+		return 0, io.EOF
+	}
+
+	// copy the buffer data into the provided byte slice
+	numberOfElementsCopied := copy(p, b.buf)
+
+	// update the buffer by removing the elements that were read
+	// (making a slice starting at the index of how many elements were read)
+	b.buf = b.buf[numberOfElementsCopied:]
+
+	// return the number of bytes read and nil error
+	return numberOfElementsCopied, nil
+}
+
+// the original Buffer Len() method signature:
+// func (b *Buffer) Len() int
+func (b *OurByteBuffer) Len() int {
+	// return the length of the buffer
+	return len(b.buf)
+}
+
+// the original bytes.NewBuffer() function signature:
+// func NewBuffer(buf []byte) *Buffer
+func NewOurBuffer(initialByteSlice []byte) *OurByteBuffer {
+	// create a new OurByteBuffer instance
+	// and populate the buffer with the initialByteSlice
+	// returning a pointer to the new OurByteBuffer instance
+	return &OurByteBuffer{buf: initialByteSlice}
+}
+
+// ---------- Implementing a custom filter
+
+// define a new struct
+// it's only field is a writer of type io.Writer
+type FilteringPipe struct {
+	writer io.Writer
+}
+
+// the original io.Writer Write() method
+// Write(p []byte) (n int, err error)
+func (fp *FilteringPipe) Write(p []byte) (int, error) {
+	// create a new slice to store the filtered bytes in
+	var filteredBytesSlice []byte
```

Comment:
If we're going to allocate a second buffer, I'd probably pre-allocate the capacity of this `[]byte` by using something like:

```go
filteredBytesSlice := make([]byte, 0, len(p))
```

Can you find out what the difference is?

Reply:
if we give it a specific length, then it doesn't grow dynamically,
so does that make it more efficient in some way?

> "As you append elements to the slice, it will not need to allocate a new underlying array and copy elements until the number of filtered bytes exceeds the pre-allocated capacity."

But I don't have a good idea about what happens under the hood with make and append